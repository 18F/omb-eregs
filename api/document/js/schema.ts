import { Node, NodeSpec, Schema } from 'prosemirror-model';

const listSchemaNodes: { [name: string]: NodeSpec } = {
  list: {
    content: 'listitem+',
    group: 'block',
    toDOM: () => ['ol', { class: 'node-list' }, 0],
  },
  listitem: {
    attrs: {
      marker: { default: '1.' },
    },
    content: 'block+',
    toDOM: node => [
      'li',
      { class: 'node-list-item' },
      ['span', { class: 'list-item-marker' }, node.attrs.marker],
      ['div', { class: 'list-item-text' }, 0],
    ],
  },
};

const schema = new Schema({
  topNode: 'policy',
  nodes: {
    policy: {
      content: 'block+',
    },
    inline: {
      content: 'text+',
      toDOM: () => ['p', { class: 'node-paragraph-text' }, 0],
    },
    heading: {
      content: 'text+',
      group: 'block',
      attrs: {
        depth: { default: 1 },
      },
      toDOM: node => [`h${node.attrs.depth}`, { class: 'node-heading' }, 0],
    },
    para: {
      content: 'inline block*',
      group: 'block',
      toDOM: () => ['div', { class: 'node-paragraph' }, 0],
    },
    sec: {
      content: 'heading block+',
      group: 'block',
      toDOM: () => ['section', { class: 'node-section' }, 0],
    },
    text: {},
    unimplementedNode: {
      group: 'block',
      atom: true,
      attrs: {
        data: {}, // will hold unrendered content
      },
      toDOM(node) {
        const nodeType = node.attrs.data.node_type || '[no-node-type]';
        return ['div', { class: 'unimplemented whatever' }, nodeType];
      },
    },
    ...listSchemaNodes,
  },
  marks: {
    external_link: {
      attrs: {
        href: {},
      },
      toDOM(data) {
        return ['a', { class: `external-link`, href: data.attrs.href }];
      },
    },
    unimplementedMark: {
      attrs: {
        data: {}, // will hold unrendered content
      },
      toDOM: data => ['span', { class: `unimplemented` }],
    },
  },
});

export const factory = {
  heading: (text: string, depth: number) =>
    schema.nodes.heading.create({ depth }, schema.text(text)),
  list: (children?: Node[]) =>
    schema.nodes.list.create({}, children || []),
  listitem: (marker: string, children?: Node[]) =>
    schema.nodes.listitem.create({ marker }, children || []),
  para: (textContent: string | Node[], children?: Node[]) =>
    schema.nodes.para.create({}, [schema.nodes.inline.create(
      {},
      typeof textContent === 'string' ? schema.text(textContent) : textContent,
    )].concat(children || [])),
  policy: (children?: Node[]) =>
    schema.nodes.policy.create({}, children || []),
  sec: (children?: Node[]) =>
    schema.nodes.sec.create({}, children || []),
  external_link: (href: string) =>
    schema.marks.external_link.create({ href }),
  unimplementedMark: (original: any) =>
    schema.marks.unimplementedMark.create({ data: original }),
  unimplementedNode: (original: any) =>
    schema.nodes.unimplementedNode.create({ data: original }),
};

export default schema;
