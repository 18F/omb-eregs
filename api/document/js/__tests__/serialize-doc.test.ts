import { Fragment } from 'prosemirror-model';

import serializeDoc, { apiFactory, convertTexts } from '../serialize-doc';
import schema, { factory } from '../schema';


describe('serializeDoc()', () => {
  it('converts nested nodes', () => {
    const node = factory.policy([
      factory.sec([
        factory.heading('Some heading', 1),
        factory.para('First paragraph'),
      ]),
      factory.para('A later paragraph'),
    ]);

    const result = serializeDoc(node);
    expect(result).toEqual(apiFactory.node('policy', {
      children: [
        apiFactory.node('sec', {
          title: 'Some heading',
          children: [
            apiFactory.node('heading', {
              content: [apiFactory.text('Some heading')],
            }),
            apiFactory.node('para', {
              content: [apiFactory.text('First paragraph')],
            }),
          ],
        }),
        apiFactory.node('para', {
          content: [apiFactory.text('A later paragraph')],
        }),
      ],
    }));
  });

  it('converts headings', () => {
    const node = factory.heading('Stuff stuff', 2); // depth is ignored
    const result = serializeDoc(node);
    expect(result).toEqual({
      node_type: 'heading',
      children: [],
      content: [apiFactory.text('Stuff stuff')],
    });
  });

  it('converts heading text to section title', () => {
    const node = factory.sec([
      factory.heading('Stufff', 1),
    ]);
    const result = serializeDoc(node);
    expect(result.title).toEqual('Stufff');
  });

  it('converts unimplemented nodes', () => {
    const node = factory.unimplementedNode({ some: 'random', attrs: 'here' });
    const result = serializeDoc(node);
    expect(result).toEqual({ some: 'random', attrs: 'here' });
  });

  it('converts list nodes', () => {
    const node = factory.list([
      factory.listitem('a)', [
        factory.para('First p'),
        factory.para('Second p'),
      ]),
      factory.listitem('2.', [factory.para('Content')]),
    ]);

    const result = serializeDoc(node);
    expect(result).toEqual(apiFactory.node('list', {
      children: [
        apiFactory.node('listitem', {
          marker: 'a)',
          type_emblem: 'a',
          children: [
            apiFactory.node('para', { content: [apiFactory.text('First p')] }),
            apiFactory.node('para', { content: [apiFactory.text('Second p')] }),
          ],
        }),
        apiFactory.node('listitem', {
          marker: '2.',
          type_emblem: '2',
          children: [
            apiFactory.node('para', { content: [apiFactory.text('Content')] }),
          ],
        }),
      ],
    }));
  });
});

describe('convertTexts()', () => {
  it('converts simple text', () => {
    const result = convertTexts(Fragment.fromArray([
      schema.text('Some content'),
    ]));
    expect(result).toEqual([{
      content_type: '__text__',
      inlines: [],
      text: 'Some content',
    }]);
  });

  it('converts nested marks', () => {
    const result = convertTexts(Fragment.fromArray([
      schema.text('Some '),
      schema.text('nested', [
        factory.unimplementedMark({ content_type: 'one', outer: 'thing' }),
        factory.unimplementedMark({ content_type: 'two', inner: 'here' }),
        factory.unimplementedMark({ content_type: 'three', most: 'deep' }),
      ]),
      schema.text(' content'),
    ]));

    expect(result).toEqual([
      apiFactory.text('Some '),
      apiFactory.content('one', {
        outer: 'thing',
        inlines: [apiFactory.content('two', {
          inner: 'here',
          inlines: [apiFactory.content('three', {
            most: 'deep',
            inlines: [apiFactory.text('nested')],
          })],
        })],
      }),
      apiFactory.text(' content'),
    ]);
  });
});
