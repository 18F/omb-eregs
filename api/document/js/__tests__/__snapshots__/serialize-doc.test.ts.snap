// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializeDoc() converts footnotes 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "content": Array [
        Object {
          "content_type": "__text__",
          "inlines": Array [],
          "text": "i am a footnote",
        },
      ],
      "marker": "2",
      "node_type": "footnote",
      "type_emblem": "2",
    },
  ],
  "content": Array [
    Object {
      "content_type": "__text__",
      "inlines": Array [],
      "text": "hello",
    },
    Object {
      "content_type": "footnote_citation",
      "inlines": Array [
        Object {
          "content_type": "__text__",
          "inlines": Array [],
          "text": "2",
        },
      ],
      "text": "",
    },
  ],
  "node_type": "para",
}
`;
